[Template] Rework motor library

rmheuer opened the pull request:
Swerve drive code is probably broken. Don't merge yet because we still need to add inverting motors and feedforward control.

rmheuer committed:
[Template] Rework motor library

rmheuer committed:
[Template] Add motor inversion

rmheuer committed:
[Template] Add feedforward calculator

rmheuer committed:
Merge branch 'main' into motor_rework

rmheuer commented:
Can now be merged

mvog2501 committed:
[Template] Added features to Bang Bang.

The bang bang calculator was missing some important features that we used on our thrower code that can tune in how the calculator should function so I added them in with getters and setters and docs.

mvog2501 committed:
[Template] Removed docs by making inputs self explanatory.

I changed all of the "vel"s and "pos"s to velocity and position because it just looks nicer to me.

mvog2501 reviewed (changes requested):
Neutral mode should probably work very similarly to the talons in that it isn't just at 0 but all the way through the neutral deadband as set by a function.

rmheuer committed:
[Template] Add motor neutral deadband

mvog2501 reviewed (approved):
These look great, this has reminded me that there is a counterpart to this function that upgrades outputs to a nominal percent. I'll add that at lunch today.

mvog2501 committed:
[Template] Added nominal output to motor.

Nominal output allows for motors to move at their lowest possible speed. This is REALLY useful in PID controlled systems because it allows it to get right to the setpoint and not just slightly away from it. This is a better alternative to kI.

mvog2501 committed:
[Template] Merge both PID calculators into one.

Both PIDCalculators were identical so I merged them into one file. We can undo this commit if the clarity of having two different types is wanted but I think this is much cleaner.

mvog2501 committed:
[Template] Added multipliers to pid and ff outputs.

Added the ability to change the ratio of pid and ff outputs for finer control. Often I will balance far heavier on the PID than the ff.

mvog2501 committed:
[Template] Avoid divide by 0 error if encoder ticks is set wrong.

Even if it didn't throw and error but instead returned Double.INFINITY, that woudn't do good things to the motor.

mvog2501 committed:
[Template] Fix rev bug.

You typo'd

mvog2501 reviewed (changes requested):
We're mostly there! Just a few more things to fix before we can go ahead and merge.
In order of importance:
1. Be able to configure the talons with a custom configuration
2. PIDCalculator should have the exact same features as PIDController (setP, getI, etc.)
3.  Spark Max's have neutral modes too!
4. Intentinal vs Unintentional wrong ticks per rotaiton in TalonMotor
5. (After merge) Make docs consistient (Periods?, Caps?)

mvog2501 committed:
[Template] Allow for more customizable neutral deadband settings.

Allowed for more library control of Neutral Deadband by removing the hardware configuration of 0.01 (Not very small)

mvog2501 committed:
[Template] Allow for legacy TalonFXMotor creation.

To allow for a custom configured TalonFX, I included the way the legacy code worked in that it would just wrap a TalonFX instead of creating a new one.

mvog2501 committed:
[Template] Added old PIDController features and legacy conversion.

mvog2501 committed:
[Template] Added more filter functionality to encoders.

mvog2501 committed:
[Template] Added neutral mode to Spark Max

mvog2501 committed:
[Template] Swap ticks per rotation to double

This allows for the weird ticks per rotation on the johnson electric's

mvog2501 committed:
[Template] Added support for custom SRX configuration.

mvog2501 reviewed (approved):
I think it's finally ready. I added all of the missing features. The formatting of the docs is a little messy but I don't really feel like fixing all of them.

mvog2501 commented:
I'm going to squash and merge at 10:00 tomorrow.
